### PR TITLE
Fix #2757 remove lightly used and unnecessary pkcollections functions

### LINKED ISSUE #2757: Remove deprecated pkcollections functions and OrderedMapping
map_items, OrderedMapping

### PR DESCRIPTION
**Pull request by robnagler**

### CODE REVIEW

File: sirepo/sim_data/warpvnd.py [hunk 1]
```diff
@@ -44,11 +44,11 @@ def _compute_model(cls, analysis_model, *args, **kwargs):
...
     def fixup_old_data(cls, data):
         dm = data.models
-        dm.pksetdefault(optimizer=PKDict())
-        dm.optimizer.setdefault(
-            constraints=[],
-            enabledFields=PKDict(),
-            fields=[],
+        dm.pksetdefault(optimizer=PKDict)
+        dm.optimizer.pksetdefault(
+            constraints=list,
```
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/sim_data/warpvnd.py:49]: This is one of the reasons for pksetdefault: Values will be called if they are callable

```py
        for k, v in i:
            if k not in self:
                self[k] = v() if callable(v) else v
```

This avoids unnecessary object creations when the default is not set.

```py
>>> from pykern import pkjson
>>> from pykern.pkcollections import PKDict
>>> print(pkjson.dump_pretty(PKDict().pksetdefault(x=[], y=list, q=list())))
{
    "q": [],
    "x": [],
    "y": []
}
```

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/warpvnd.py:49]: Make sense. Thanks.

Should I add this to our coding style notes? 

One concern is this seem like writing extra code to optimize for a problem I didn't know we had. Do you feel it's worth it?

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/sim_data/warpvnd.py:49]: It was created for the complex case where you don't want to instantiate something when there is already an instance, e.g. 
[get_instance](https://github.com/radiasoft/sirepo/blob/4a967584b273c4c922c38c2cfcfeab4aaea569fe/sirepo/job_driver/sbatch.py#L58):
```py
    def get_instance(cls, op):
        u = op.msg.uid
        return cls.__instances.pksetdefault(u, lambda: cls(op))[u]
```

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/warpvnd.py:49]: Sorry I wasn't very clear. I get the use case generally but in this specific instance we are just instantiating an empty list which doesn't seem like much of a problem.

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/sim_data/warpvnd.py:49]: I agree it doesn't matter, but I was editing the code to fix its case (passing kwargs to pksetdefault, which setdefault doesn't support), and I made the change at the same time. I also changed `x.update(PKDict(a=1, b=2))` to `x.update(a=1, b=2)`, because the PKDict call is unnecessary but not wrote (@moellep take note).

The pros and cons are interesting: People who know setdefault will not know what pksetdefault is or does. The whole PKDict thing is an unnecessary optimization. Pythonistas "en masse" have not invented it therefore it is by definition unnecessary.

However, I think PKDict is useful. I also think the model of passing a pksetdefault is useful, having used it for a couple of decades in Perl. Therefore, I think demonstrating it in the code is useful to the team. As you know, my code is not exactly understandable/well-liked by Pythonistas (or Perlers for that matter :).

There is a cognitive load that you ran into when you saw it, which caused this thread, but I guarantee you that you will away with the knowledge of what pksetdefault does, now. That's a good thing. Cognitive load that improves skills.

The actual pattern is Lazy Initialization, and people write [epics](https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization) about it, and even [apply YAGNI to it](https://stackoverflow.com/a/978811), which is funny to me, since you are saying YAGNI to the use here. The interesting part is that nobody seems to solve it by making the properties with a PKDict.pksetdefault style, which clearly saves code.

As far as performance goes, it's a non-issue, of course, but I was curious so I ran a test on my VM. `list()` takes 90ns and `[]` takes 9ns, which would argue for not calling `list` and instead using `[]` syntax. Curiosity got the best of me. `{}` takes 16ns and `dict()` takes 100ns. My change is definitely not a performance optimization. And, this is one advantage of "syntax": the compiler can inline operations that can't be overridden dynamically. `list` must be looked up and dispatched whereas `[]` goes directly to the bytecode `BUILDLIST`:

```sh
$ python -c 'import dis; dis.dis(lambda: [])'
  1           0 BUILD_LIST               0
              2 RETURN_VALUE
```
whereas `list()` generates a function load variable and a call:

```sh
$ python -c 'import dis; dis.dis(lambda: list())'
  1           0 LOAD_GLOBAL              0 (list)
              2 CALL_FUNCTION            0
              4 RETURN_VALUE
```

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/warpvnd.py:49]: As always thanks. You're right to say I won't be forgetting how pksetdefault works anytime soon;)

[] vs list() speed is something I've never thought about before but makes sense in a dynamic language like Python.

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/sim_data/warpvnd.py:49]: > Should I add this to our coding style notes?

No.

TL;DR I don't know if this is a style issue as how to use pksetdefault. It kind of depends on particular context. In this context, it doesn't matter, because it is a fixup, which runs lazily. 

If we were using it in an environment where the default is rarely set, then passing the function is cheaper and better for AGC. OTOH, if the default is normally set, e.g. [job_api._request_content](https://git.radiasoft.org/sirepo/blob/4a967584b273c4c922c38c2cfcfeab4aaea569fe/sirepo/job_api.py#L229), then passing an already created empty object is probably better. However, I think the uses of lambda are appropriate in _request_content. It also sets a good example of 1) allow callers to override certain values (and not others and 2) using a feature that is not seen much. I just fixed [job_cmd._background_percent_complete](https://git.radiasoft.org/sirepo/blob/4a967584b273c4c922c38c2cfcfeab4aaea569fe/sirepo/pkcli/job_cmd.py#L69) to do [the right thing](https://git.radiasoft.org/sirepo/pull/2758/commits/b7956dd736b7ff5aa3d9e069eeabe45bc9014a21).

Timing operations is less important these days, but I think it's useful to understand that nothing is free, and "good habits" make for faster code, in general. And, often, makes for more reliable code, but obviously sometimes these things are at odds. 

I don't think there's a way to document this. I continually point people to [Hints on Computer System Design (1983)](http://n99.us/mob), because it talks about a lot of issues like this. If it were that easy, Lampson would not have to talk about the [same things in 2013](http://n99.us/pyh). 

Aside but relevant, perhaps... Consider the *very* simple problem of keeping URLs around, permanently. I went to both of these shortlinks, and they both were broken. I had to update them. That's why I use shortlinks to articles like this, because it is good information architecture. However, how many people use shortlinks for this reason? How many people use it for the wrong reason? And, why is that Microsoft and the Heidelberg Laureate Forum Foundation can't keep a link stable? [Link rot](https://en.wikipedia.org/wiki/Link_rot) is documented, well-studied. Good habits are hard to establish. Documentation alone will not do this.

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/warpvnd.py:49]: Ok. I don’t agree but I’m happy to write a personal blog post on this style instead.

My reasoning:

Using pksetdefault itself is a style issue. That is our style. Using list instead of list() or [] in pksetdefault is a style issue. Everywhere else where we define a list we do it like [] so when I look at this code I think, why? This should be documented. Yes it is “how it works” but optimizing these use cases is now our style. That’s new to me and I’d like to document it so I can remember it. Our code has evolved over time and so has the style. It would be nice to document the latest style (at the very least for me and I’m willing to take on the work of documenting it). When I look across our code base we have areas that haven’t been touched in a while. I don’t have the “good taste” to know what is the current style so I like to have something I can point to. My current thoughts are “what would Rob think” which is poorly defined. In the pykern wiki you say: “The most important one is that we didn't have to think much about how to build applications: policies were defined, and we followed them. If policies needed changing, they were changed, and we all followed the new ones.” As you know I like things like this, preferring convention over configuration. It reduces my cognitive load and frees me to think about all of the other hard problems I’m struggling to solve. By having a coding style document and __using it__ we establish shared and agreed upon style. If we don’t update it with the latest style then the styles are poorly enforced. This weak enforcement forces me to think about the problem each time I encounter it instead of knowing what the style is and refactoring the outdated code. You can refactor because you have the taste and knowledge that I lack. Documenting style would help me gain taste.

[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/sim_data/warpvnd.py:49]: Agreed. I was mostly saying where it would be documented. :) I think the use of pksetdefault is documentation in pkcollections.py, which can be referred to in the style guide. In particular, I would note that `x=list` if slower than `x=[]`, which is non-obvious, but someone looking at pksetdefault documentation should know.

The use of PKDict over dict is a style preference, especially as a base class. That's an important design decision, actually.

Another thing to document is use `[]` instead of `list()` for empty lists, because it is faster and there's no particular difference between the two. 

Use `PKDict()` instead of `{}`, even though slower, it's more convenient.

There are other style issues `PKDict(a=1)` is preferred over `PKDict({'a': 1})`, because even though the latter may be faster (not tested). Depends on whether CALL_FUNCTION_KW is faster than BUILD_MAP, CALL_FUNCTION.

Documenting these things is definitely style.

[the remark this answers is not in the document]
Comment (e-carlin) [reply sirepo/sim_data/warpvnd.py:49]: Ah, sorry I missed that:). I'll add what you say here and my thoughts to pkcollections/design hints.
